### ISSUE TITLE
Reconstructing types defined in one module inside another

### ISSUE THREAD

**Issue by qua4tre** [JuliaIO/JLD2.jl#107]
The following worked in julia v0.6 but no longer works in v0.7:

I have a type defined in ModA.jl (in `LOAD_PATH`):
```
module ModA

struct MyType
    a::Int64
    b::Float64
end

end # module
```

I save data with my type in a JLD2 file:
```
julia> import ModA
[ Info: Precompiling ModA [top-level]

julia> using FileIO

julia> t = ModA.MyType(1, 0.0)
ModA.MyType(1, 0.0)

julia> save("t.jld2", "t", t)
```

In some code I run separately, I try to load the data inside a module and I get a warning:
```
julia> module ModB

       import ModA
       using FileIO

       t = load("t.jld2", "t")

       end
┌ Warning: type ModA.MyType does not exist in workspace; reconstructing
└ @ JLD2 ~/.julia/packages/JLD2/xpVxm/src/data.jl:1153
Main.ModB

julia> ModB.t
getfield(JLD2.ReconstructedTypes, Symbol("##ModA.MyType#368"))(1, 0.0)
```

Then if I try to interact with `ModB.t` (e.g. call functions on it defined only on type `MyType`), I get an error (because the type is wrong).  This is similar to #82; may be related?

The problem is alleviated by doing an `import ModA` at the top level before I try to load `ModB`, so I suspect the reason this worked on v0.6 and doesn't on v0.7 is due to (https://github.com/JuliaLang/julia/issues/17997).  I'm wondering if this just means I need to change my workflow and always load modules with user-defined types at the top level before trying to load files containing those types, or if this is something that can be addressed in JLD2?  For example, would there be a way to make JLD2 "see" that `ModA` has been `import`ed inside `ModB` and recognize `ModA.MyType` without the need to explicitly `import ModA` at the top level?

**Comment by jingpengw** [comment]
I got a similar issue. this blocked my work!

**Comment by floswald** [comment]
same here.

```
julia> ev
Eval Object:
============

Objective function value: -1.0
Evaluation Time: 1.536930563524261e9
Evaluation Status: 1
Parameters:
Any[:a, :b]
Moments:
Symbol[:alpha, :beta, :gamma]
===========================

julia> fieldnames(typeof(ev))
(:value, :time, :params, :simMoments, :dataMoments, :dataMomentsW, :status, :prob, :accepted, :options)

julia> MomentOpt.FileIO.save("test5.jld2",Dict("ev"=>ev))

julia> MomentOpt.FileIO.load("test5.jld2", "ev")
┌ Warning: type DataStructures.OrderedDict{Any,Any} does not exist in workspace; reconstructing
└ @ JLD2 ~/.julia/packages/JLD2/KjBIK/src/data.jl:1153
Error encountered while loading "/Users/74097/.julia/v0.6/MomentOpt/test5.jld2".
Fatal error:
ERROR: MethodError: Cannot `convert` an object of type getfield(JLD2.ReconstructedTypes, Symbol("##DataStructures.OrderedDict{Any,Any}#370")) to an object of type DataStructures.OrderedDict
Closest candidates are:
  convert(::Type{T<:AbstractDict}, ::T<:AbstractDict) where T<:AbstractDict at abstractdict.jl:513
  convert(::Type{T<:AbstractDict}, ::AbstractDict) where T<:AbstractDict at abstractdict.jl:516
  convert(::Type{S}, ::T<:(Union{CategoricalString{R}, CategoricalValue{T,R} where T} where R)) where {S, T<:(Union{CategoricalString{R}, CategoricalValue{T,R} where T} where R)} at /Users/74097/.julia/packages/CategoricalArrays/04bks/src/value.jl:66
  ...
Stacktrace:
 [1] jlconvert at /Users/74097/.julia/packages/JLD2/KjBIK/src/data.jl:675 [inlined]
 [2] macro expansion at /Users/74097/.julia/packages/JLD2/KjBIK/src/data.jl:1285 [inlined]
 [3] jlconvert(::JLD2.ReadRepresentation{Eval,JLD2.OnDiskRepresentation{(0, 8, 16, 24, 32, 40, 48, 56, 64, 65),Tuple{Float64,Float64,DataStructures.OrderedDict,DataStructures.OrderedDict,DataStructures.OrderedDict,DataStructures.OrderedDict,Int64,Float64,Bool,Dict},Tuple{Float64,Float64,JLD2.RelOffset,JLD2.RelOffset,JLD2.RelOffset,JLD2.RelOffset,Int64,Float64,Bool,JLD2.RelOffset}}()}, ::JLD2.JLDFile{JLD2.MmapIO}, ::Ptr{Nothing}, ::JLD2.RelOffset) at /Users/74097/.julia/packages/JLD2/KjBIK/src/data.jl:1232
 [4] read_data(::JLD2.JLDFile{JLD2.MmapIO}, ::JLD2.ReadRepresentation{Eval,JLD2.OnDiskRepresentation{(0, 8, 16, 24, 32, 40, 48, 56, 64, 65),Tuple{Float64,Float64,DataStructures.OrderedDict,DataStructures.OrderedDict,DataStructures.OrderedDict,DataStructures.OrderedDict,Int64,Float64,Bool,Dict},Tuple{Float64,Float64,JLD2.RelOffset,JLD2.RelOffset,JLD2.RelOffset,JLD2.RelOffset,Int64,Float64,Bool,JLD2.RelOffset}}()}, ::Array{JLD2.ReadAttribute,1}) at /Users/74097/.julia/packages/JLD2/KjBIK/src/dataio.jl:37
 [5] read_data(::JLD2.JLDFile{JLD2.MmapIO}, ::JLD2.ReadDataspace, ::UInt8, ::Int64, ::Int64, ::Int64, ::UInt16, ::JLD2.RelOffset, ::Array{JLD2.ReadAttribute,1}) at /Users/74097/.julia/packages/JLD2/KjBIK/src/datasets.jl:149
 [6] load_dataset(::JLD2.JLDFile{JLD2.MmapIO}, ::JLD2.RelOffset) at /Users/74097/.julia/packages/JLD2/KjBIK/src/datasets.jl:92
 [7] getindex(::JLD2.Group{JLD2.JLDFile{JLD2.MmapIO}}, ::String) at /Users/74097/.julia/packages/JLD2/KjBIK/src/groups.jl:108
 [8] read(::JLD2.JLDFile{JLD2.MmapIO}, ::String) at /Users/74097/.julia/packages/JLD2/KjBIK/src/JLD2.jl:326
 [9] #45 at /Users/74097/.julia/packages/JLD2/KjBIK/src/loadsave.jl:134 [inlined]
 [10] #jldopen#31(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::getfield(JLD2, Symbol("##45#46")){String}, ::String, ::Vararg{String,N} where N) at /Users/74097/.julia/packages/JLD2/KjBIK/src/loadsave.jl:4
 [11] jldopen at /Users/74097/.julia/packages/JLD2/KjBIK/src/loadsave.jl:2 [inlined]
 [12] #load#44 at /Users/74097/.julia/packages/JLD2/KjBIK/src/loadsave.jl:133 [inlined]
 [13] load(::FileIO.File{FileIO.DataFormat{:JLD2}}, ::String) at /Users/74097/.julia/packages/JLD2/KjBIK/src/loadsave.jl:133
 [14] top-level scope at none:0
 [15] eval at ./boot.jl:319 [inlined]
 [16] #load#27(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::FileIO.File{FileIO.DataFormat{:JLD2}}, ::String) at /Users/74097/.julia/packages/FileIO/kflOT/src/loadsave.jl:188
 [17] load(::FileIO.File{FileIO.DataFormat{:JLD2}}, ::String) at /Users/74097/.julia/packages/FileIO/kflOT/src/loadsave.jl:172
 [18] #load#13(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::String, ::String) at /Users/74097/.julia/packages/FileIO/kflOT/src/loadsave.jl:118
 [19] load(::String, ::String) at /Users/74097/.julia/packages/FileIO/kflOT/src/loadsave.jl:118
 [20] top-level scope at none:0

```

**Comment by MoonCoral** [comment]
This screwed me over to.
The behaviour I would personally like to see is:
- to first try to to import missing modules
- failing that not to blindly create a new data type but to store the data in a JLD2 type such 
  -  that if it is written back to a jld2 file as if it had been properly loaded
     - that is, no risk of loss or corruption from a load save cycle
     - a use case would be to concatenate two jld2 files produced by a distributed process
  - that the data can be manually recovered in an interactive environment, similar to the current environment

**Comment by BoundaryValueProblems** [comment]
Yes, I have the same problem! Interestingly, the current version of JLD v0.9.0 works without any problem of saving and loading such composite data types under Julia v1.0.1 although the saved file size is larger than that of JLD2.

**Comment by richiejp** [comment]
This might be related to #126 

You can call `JLD2.typename` on a type to see what fully qualified name is being used.

**Comment by kleinschmidt** [comment]
This happens for me even for types defined in packages that are on the load path/project:

```
julia> module Sandbox3
           using DataFrames, JLD2
           x = DataFrame(a = 1:10)
           @save "x.jld2" x
           @load "x.jld2"
       end
[ Info: Recompiling stale cache file /home/dave/.julia/compiled/v1.0/JLD2/O1EyT.ji for JLD2 [033835bb-8acc-5ee8-8aae-3f567f8a3819]
┌ Warning: type DataFrames.DataFrame does not exist in workspace; reconstructing
└ @ JLD2 ~/.julia/dev/JLD2/src/data.jl:1143
┌ Warning: type DataFrames.Index does not exist in workspace; reconstructing
└ @ JLD2 ~/.julia/dev/JLD2/src/data.jl:1143
Main.Sandbox3
```

**Comment by kleinschmidt** [comment]
Which makes it basically impossible to use JLD2 in Weave.jl (at least without turning sandboxing off which `@eval`s all the code in `Main`) (mpastell/Weave.jl#164)

**Comment by photicus** [comment]
I'm seeing the same issue in trying to upgrade some code from Julia 0.6. What seems to work is to have all the necessary packages loaded in the `Main` module. When I run what @kleinschmidt in the REPL I see the same issue. However, if I put in `using DataFrames` before defining the module it works just fine. I'm note sure why this works though...

This isn't a viable workaround in the long run, but hopefully it sheds some light on what's going on.

**Comment by qua4tre** [comment]
It seems JLD2 looks for modules in `Main` only and can't see modules `import`ed inside the current module.

A related workaround I found (can't remember where right now) is to add a line `Core.eval(Main, :(import MyMod))` for each module that defines types that I need JLD2 to be able to recognize.

**Comment by photicus** [comment]
So I tried seeing what happens if I tried using JLD instead of JLD2. The code above still performes a reconstruction warning (not the case in Julia0.6). However, with JLD there's the `addrequire` function which provides automatic loading of modules into `Main`. I don't know if there's something similar for JLD2.

**Comment by AmebaBrain** [comment]
> It seems JLD2 looks for modules in `Main` only and can't see modules `import`ed inside the current module.
> 
> A related workaround I found (can't remember where right now) is to add a line `Core.eval(Main, :(import MyMod))` for each module that defines types that I need JLD2 to be able to recognize.

Probably it was [here](https://github.com/JuliaIO/JLD.jl/issues/216#issuecomment-431693169)  :)

Adding `Core.eval(Main, :(using SomeModule))` into your module `__init__` function allows you to automatically push all required modules definition into the Main module scope during execution of `using MyModule` and thus make issue resolved

**Comment by timholy** [comment]
This should have been fixed by #151, #164, #166, and #173, which are part of 0.1.11. If you're using this version and still encountering problems, please file a new issue.

**Comment by pjuangph** [comment]
I am still seeing this problem I have the same set up as OP.
